plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java

```diff
@@ -0,0 +1,179 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.repositories.gcs;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.core.internal.io.IOUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+/**
+ * Wrapper around reads from GCS that will retry blob downloads that fail part-way through, resuming from where the failure occurred.
+ * This should be handled by the SDK but it isn't today. This should be revisited in the future (e.g. before removing
+ * the {@link org.elasticsearch.Version#V_7_0_0} version constant) and removed if the SDK handles retries itself in the future.
+ */
+class GoogleCloudStorageRetryingInputStream extends InputStream {
+
+    private static final Logger logger = LogManager.getLogger(GoogleCloudStorageRetryingInputStream.class);
+
+    static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private final Storage client;
+
+    private final BlobId blobId;
+
+    private final int maxRetries;
+
+    private InputStream currentStream;
+    private int attempt = 1;
+    private List<StorageException> failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
+    private long currentOffset;
+    private boolean closed;
+
+    GoogleCloudStorageRetryingInputStream(Storage client, BlobId blobId) throws IOException {
+        this.client = client;
+        this.blobId = blobId;
+        this.maxRetries = client.getOptions().getRetrySettings().getMaxAttempts() + 1;
+        currentStream = openStream();
+    }
+
+    private InputStream openStream() throws IOException {
+        try {
+            final ReadChannel readChannel = SocketAccess.doPrivilegedIOException(() -> client.reader(blobId));
+            readChannel.seek(currentOffset);
```

Comment:
Maybe only seek if `currentOffset > 0L`?

Reply:
This doesn't have any IO impact in the SDK (it just sets an long field that is then used when firing off the actual request) so I figured why do 3 lines instead of 1 unless we have to? :)

Reply:
> I figured why do 3 lines instead of 1 unless we have to? :)

I see things differently, why always seek to the beginning when most of the time it's not necessary? :)

Anyway, it's nitpicking so do as you prefer :)